### PR TITLE
Using Result<,> instead of Choice<_,'Error>

### PR DESCRIPTION
**Pull request by wallymathieu**
- Changed to Result in order to follow the FSharp semantics
- Added `Result.ofChoice` and `Result.toChoice`

### CODE REVIEW

File: ExtCore.Tests/Control.fs [hunk 1]
```diff
@@ -38,7 +38,7 @@ type AsyncChoiceBuilderDisposeFixture() =
                 disposed.Value <- true }
 
     let createAsyncChoiceDisposable() =
```
Comment (vasily-kirichenko) [inline ExtCore.Tests/Control.fs]: I think all the functions which have "Choice" word in their names should be renamed.

Comment (wallymathieu) [reply ExtCore.Tests/Control.fs]: I agree

File: ExtCore/Pervasive.fs [hunk 2]
```diff
@@ -861,6 +850,17 @@ module Option =
...
+        | Ok result ->
+            Some result
+        | Error _ ->
+            None
+
+    /// <summary></summary>
+    /// <param name="value"></param>
+    /// <returns></returns>
     [<CompiledName("ToChoice")>]
     let toChoice (value : 'T option) : Choice<'T, unit> =
```
Comment (vasily-kirichenko) [inline ExtCore/Pervasive.fs]: Maybe just remove it?

File: ExtCore/Pervasive.fs [hunk 3]
```diff
@@ -1012,188 +1034,141 @@ module Option =
...
+        | Error _ -> false
 
-    /// <summary>Does the Choice value represent an error value?</summary>
+    /// <summary>Does the Result value represent an error value?</summary>
     /// <param name="value"></param>
     /// <returns></returns>
     [<CompiledName("IsError")>]
-    let inline isError (value : Choice<'T, 'Error>) : bool =
-        // Preconditions
-        checkNonNull "value" value
```
Comment (vasily-kirichenko) [inline ExtCore/Pervasive.fs]: Yeah, eliminating null check is great :)

File: ExtCore/Pervasive.fs [hunk 4]
```diff
@@ -1012,188 +1034,141 @@ module Option =
...
+            Ok result
         | None ->
-            Choice2Of2 ()
+            Error ()
+
+    /// <summary></summary>
+    /// <param name="value"></param>
+    /// <returns></returns>
+    [<CompiledName("OfChoice")>]
+    let ofChoice (value : Choice<'T,'Error>) : Result<'T, 'Error> =
```
Comment (vasily-kirichenko) [inline ExtCore/Pervasive.fs]: Is it needed?

File: ExtCore/Pervasive.fs [hunk 5]
```diff
@@ -1012,188 +1034,141 @@ module Option =
...
+    /// <summary></summary>
     /// <param name="value"></param>
     /// <returns></returns>
-    [<CompiledName("MapError")>]
-    let mapError (mapping : 'Error1 -> 'Error2) (value : Choice<'T, 'Error1>) =
-        // Preconditions
-        checkNonNull "value" value
-
+    [<CompiledName("ToChoice")>]
+    let toChoice (value : Result<'T, 'Error>) : Choice<'T,'Error> =
```
Comment (vasily-kirichenko) [inline ExtCore/Pervasive.fs]: Same, do we need it?

File: ExtCore/Pervasive.fs [hunk 6]
```diff
@@ -1267,65 +1230,53 @@ module Choice =
...
 
     /// <summary></summary>
     /// <param name="value"></param>
     /// <returns></returns>
     [<CompiledName("BindOrFail")>]
-    let inline bindOrFail (value : Choice<'T, string>) : 'T =
-        // Preconditions
-        checkNonNull "value" value
-
+    let inline bindOrFail (value : Result<'T, string>) : 'T =
```
Comment (vasily-kirichenko) [inline ExtCore/Pervasive.fs]: Maybe `bindOrError`?

**Comment by vasily-kirichenko** [comment]
This is awesome.

File: ExtCore/Pervasive.fs (hunk 6 above)
Comment (wallymathieu) [reply ExtCore/Pervasive.fs]: Well, in this case it does a `fail` so it does not really return an error but raises an exception.

Comment (vasily-kirichenko) [reply ExtCore/Pervasive.fs]: Ah, sorry :)

File: ExtCore/Pervasive.fs (hunk 5 above)
Comment (wallymathieu) [reply ExtCore/Pervasive.fs]: In order to be compatible with libraries that does not implement `Result` ?

File: ExtCore/Pervasive.fs (hunk 6 above)
Comment (wallymathieu) [reply ExtCore/Pervasive.fs]: No worries :)

**Comment by wallymathieu** [comment]
I'm trying to debug what goes wrong with the Patricia hash map, but from what I can see it's not ordering itself in the same way from time to time.
Sample outputs are:
```
"..2...3....4.1right"
"...3..2.1....4right"
".1....4...3..2right"
```
For the fold test there is no assumption of ordering, only that the application is done. 
```
        // reference keys
        let refMap = HashMap.ofSeq [for c in ["."; ".."; "..."; "...."] do yield (c, c.Length) ]
        let resultRefMap = refMap |> HashMap.fold  (fun x y z -> (y,z) :: x)  [("initial",83)]
        CollectionAssert.AreEquivalent([("initial",83);("....",4);("...",3);("..",2);(".",1)], resultRefMap)
        Assert.AreEqual(("initial",83), resultRefMap |> Array.ofSeq |> Array.last)
```
so a naive porting of that to foldback would be:
```
        // reference keys
        let refMap = HashMap.ofSeq [for c in ["."; ".."; "..."; "...."] do yield (c, c.Length) ]
        let resultRefMap =  HashMap.foldBack  (fun x y z -> (x,y) :: z) refMap [("initial",83)]
        CollectionAssert.AreEquivalent([("initial",83);("....",4);("...",3);("..",2);(".",1)], resultRefMap)
        Assert.AreEqual(("initial",83), resultRefMap |> Array.ofSeq |> Array.last)
```
that does look like it succeeds locally

**Comment by vasily-kirichenko** [comment]
@jack-pappas What do you think about this? I'm for merging it and publish as Nuget package version 1.0.0.

**Comment by vasily-kirichenko** [comment]
@jack-pappas @wallymathieu I create a vote on twitter https://twitter.com/kot_2010/status/924710700597080064, let's see what people think.

**Comment by jack-pappas** [comment]
I'll take a look over the changes. My initial thought is that this should be an addition to (instead of replacement for) the ``Choice<_,_>``-based functions. I think it's best if ExtCore still supports (at least for a while) older versions of F# where reasonable.

Maybe what we should be doing with ExtCore is to have separate packages targeting different versions of F#? That'd let us support older versions of F# while still letting us target features available in newer versions of the language. If we did this, I think I'd still be in favor of making these changes an addition instead of a replacement -- that'd make it easier for projects using ExtCore with older versions of VS / F# to upgrade to a newer version of VS / F# (they can start using the package for the new F# version, but without having to immediately switch all of their code over to using ``Result<_,_>``.

**Comment by jack-pappas** [comment]
@wallymathieu I'm not sure that ``HashMap`` test is sensible anyway, since any ordering is going to depend on the exact specifics of the hashing function used; the string hashing function differs between .NET versions / Mono (and I think .NET Framework even has a switch to allow the string hashes to be randomized with a per-process random seed). I'll just disable or delete the tests.

**Comment by wallymathieu** [comment]
The main con of that approach is that ExtCore is not really compatible with using Result due to for instance the active pattern for Choice. One way to mitigate that would be to create a `ExtCore.Compability` namespace with the Choice active pattern et.c. (so that you can import that namespace in order to still use Choice in your code).

**Comment by wallymathieu** [comment]
I could close this pull request and create a new one with both Choice and Result?

**Comment by jack-pappas** [comment]
@wallymathieu I think that’ll be the best log-term approach. I like your idea of moving the existing Result active pattern in ExtCore to a separate, optional namespace for compatibility.

**Comment by jack-pappas** [comment]
Also, I’ll disable/delete the broken HashMap unit tests tonight so they don’t keep making noise.

**Comment by wallymathieu** [comment]
@jack-pappas I've opened a third pull request, so I'm closing this one.

**Pull request closed without merging**